File: 算法基础/1/ex1/src/1.py
import sys


def printOptimalParens(s, i, j):
    if i == j:
        print(' A' + str(i), end='')
    else:
        print('(', end='')
        printOptimalParens(s, i, s[i][j])
        printOptimalParens(s, s[i][j] + 1, j)
        print(')', end='')


def matrixChainOrder(p):
    n = len(p) - 1
    m = [[0 for i in range(n + 1)] for j in range(n + 1)]
    s = [[0 for i in range(n + 1)] for j in range(n + 1)]
    for i in range(1, n + 1):
        m[i][i] = 0
    for l in range(2, n + 1):
        for i in range(1, n - l + 2):
            j = i + l - 1
            m[i][j] = float("inf")
            for k in range(i, j):
                q = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j]
                if q < m[i][j]:
                    m[i][j] = q
                    s[i][j] = k
    return m, s


def main(epoch):
    n = int(input())
    raw = input().strip(' ').split(' ')
    p = [int(i) for i in raw]
    # n = 6
    # p = [30, 35, 15, 5, 10, 20, 25]
    for i in range(epoch):
        m, s = matrixChainOrder(p)
    print(m[1][n])
    printOptimalParens(s, 1, n)
    print()


if len(sys.argv) == 1:
    main(1)
else:
    main(int(sys.argv[1]))

File: 算法基础/1/ex1/src/1.cpp
#include <iostream>
#include <limits>

#define N 30
#define FOR(i, l, r) for (i = l; i <= r; i++)

using namespace std;

void printOptimalParens(int s[N][N], int i, int j) {
    if (i == j)
        cout << " A" << to_string(i);
    else {
        cout << '(';
        printOptimalParens(s, i, s[i][j]);
        printOptimalParens(s, s[i][j] + 1, j);
        cout << ')';
    }
}

void matrixChainOrder(long long p[N], int n, long long m[N][N], int s[N][N]) {
    int l, i, j, k;
    long long q;
    FOR (i, 1, n)
        m[i][i] = 0;
    FOR (l, 2, n)
        FOR (i, 1, n) {
            j = i + l - 1;
            m[i][j] = numeric_limits<long long>::max();
            FOR (k, i, j - 1) {
                q = m[i][k] + m[k + 1][j] + p[i - 1] * p[k] * p[j];
                if (q < m[i][j]) {
                    m[i][j] = q;
                    s[i][j] = k;
                }
            }
        }
}

void printMS(long long m[N][N], int s[N][N], int n) {
    int i, j;
    cout.setf(ios::left);
    FOR (i, 1, n) {
        FOR (j, 1, n) {
            cout.width(16);
            cout << m[i][j];
        }
        cout << '\n';
    }
    cout << '\n';
    FOR (i, 1, n) {
        FOR (j, 1, n) {
            cout.width(16);
            cout << s[i][j];
        }
        cout << '\n';
    }
    cout << '\n';
}

int main(int argc, char* argv[]) {
    int n, i, j, epoch;
    long long p[N];
    long long m[N][N];
    int s[N][N];

    cin >> n;
    FOR (i, 0, n)
        cin >> p[i];

    if (argc > 1)
        epoch = stoi(argv[1]);
    else
        epoch = 1;
    FOR (i, 1, epoch)
        matrixChainOrder(p, n, m, s);

    if (argc > 2)
        printMS(m, s, n);

    cout << m[1][n] << '\n';
    printOptimalParens(s, 1, n);
    cout << '\n';
    return 0;
}
